### PR TITLE
pkg: add slices and maps package

### PR DESCRIPTION
**Pull request by howardjohn**
This PR introduces two new packages, `slices` and `maps`.

The motivation behind this is that there are a variety of slice/map operations that are useful in Istio but not in exp/x/slices or exp/x/maps. We could just make a package only for those additions, but then we run into needing to differentiate like istioslices.Sort vs slices.Sort. So we import all the functions we need, and modify / add as needed.

Changes from upstream ones are:
* Filter, Map, and Convert
* Contains/Find instead of Index
* Sort functions return the array for easier usage

The benefits of using these functions instead of upstream one is less boilerplate, clearer intent, and easier to optimize (for example, we do slow filtering in many places - easy to do the right thing when its just a function call away)

This PR moves a few call sites to the new package to show example usages, but it is intentionally not changing all of them since that would cause a lot of churn

### CODE REVIEW

**Comment by zirain** [comment]
there's one thing I want to ask long before:
will these package(include `sets`) move to `istio/pkg`?

**Comment by howardjohn** [comment]
Istio/pkg was formed when we had plan to have multi-repo and needed common code. Given we have consolidated on a mostly monorepo I don't think it provides much value separated and makes it harder to maintain so I wouldn't be in favor.

it's probably reasonable to fold the whole repo into Istio/Istio but not sure if everyone else is on board with that.

**Comment by jacob-delgado** [comment]
I'd be in favor of moving istio/pkg to istio. There have been times when it has needed updates or has changed and has broken istio/istio. Minimizing the chances of this would be great.

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
Most functions are a wrapper and widely used. Only function like `Filter` is added for special case. 

Not sure whether it makes much sense

**Comment by howardjohn** [comment]
there is also sort returning the slice to avoid needing multiple lines,
filter, etc. We can add more as needed (I actually have multiple WIP PR
where I added another one to help the implementation), was just keeping
this PR pretty small

**Comment by hzxuzhonghu** [comment]
Mostly i think those new functions are not very widely used. If so, we can add to upstream.

**Comment by howardjohn** [comment]
@hzxuzhonghu here is an example of a PR adding low-performance code because we do not have these abstractions: https://github.com/istio/istio/pull/44891/files#diff-e3b667b806175fc66cb679217641e1638639507c364df2432a60c1d4e87a9739R150-R154

This could have been `slices.Filter(deletedClusters, builtClusters.Contains)`, saving 7 LOC, more clear, and fewer allocations (filter in place instead of new slice)

File: pkg/slices/slices.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	return slices.Delete(s, i, i+1)
+}
+
+// Contains reports whether v is present in s.
+func Contains[E comparable](s []E, v E) bool {
+	return slices.Contains(s, v)
+}
+
+// FindFunc finds the first element matching the function, or nil if none do
+func FindFunc[E any](s []E, f func(E) bool) *E {
```
Comment (hzxuzhonghu) [inline pkg/slices/slices.go:64]: What if it is a []*pointer, but with an elment  == nil.

File: pkg/slices/slices.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	if idx == -1 {
+		return nil
+	}
+	return &s[idx]
+}
+
+// Filter retains all elements in []E that f(E) returns true for.
+// The array is *mutated in place* and returned.
+// Used Filtered to avoid mutation
+func Filter[E any](s []E, f func(E) bool) []E {
```
Comment (hzxuzhonghu) [inline pkg/slices/slices.go:75]: It is kind of confusing with `Filtered`

**Review by hzxuzhonghu — COMMENTED** [review]
Seems these wrapper are inspired by rustlang.

**Comment by howardjohn** [comment]
> Seems these wrapper are inspired by rustlang.

Filter on a slice is common in almost every language - rust, python, java, etc

File: pkg/slices/slices.go (hunk 1 above)
Comment (howardjohn) [reply pkg/slices/slices.go:64]: In that case, `E=*pointer`. So `**E` is the response.

If we made it `func FindFunc[E any](s []E, f func(E) bool) E` then you would not be able to distinguish in your case.

We could make it `func FindFunc[E any](s []E, f func(E) bool) (E, bool)` but I think its easier to use as written here

File: pkg/slices/slices.go (hunk 2 above)
Comment (howardjohn) [reply pkg/slices/slices.go:75]: I renamed it

**Comment by howardjohn** [comment]
I have built https://github.com/istio/istio/pull/45058 on top of this PR, so it would be good to get direction one way or another on this. I can remove some functions if they are controversial

File: pkg/slices/slices.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	idx := slices.IndexFunc(s, f)
+	if idx == -1 {
+		return nil
+	}
+	return &s[idx]
+}
+
+// FilterInPlace retains all elements in []E that f(E) returns true for.
+// The array is *mutated in place* and returned.
+// Used Filter to avoid mutation
```
Comment (hzxuzhonghu) [inline pkg/slices/slices.go:74]:
```suggestion
// Use Filter to avoid mutation
```

**Review by hzxuzhonghu — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

**Review by ericvn — APPROVED** [review]

**Pull request merged**